### DISCUSSION TITLE
`MenuButton` does not support using `accessibleWhenDisabled` and `render` props together

### DISCUSSION THREAD

**Discussion by mayank99** [ariakit/ariakit#4131] [General]
Consider this snippet:

```jsx
<Ariakit.MenuButton
  accessibleWhenDisabled
  disabled
/>
```

This correctly uses `aria-disabled`, allows the button to be focused, while preventing the menu from being opened in any way (clicks, <kbd>Enter</kbd>, <kbd>Spacebar</kbd>, arrow-keys).

Now let's say the `disabled` prop is set via the `render` prop. Ideally, I'd be able to do this:

```jsx
<Ariakit.MenuButton
  accessibleWhenDisabled
  render={<Ariakit.Button disabled />}
/>
```

However, that does not work at all. `<MenuButton accessibleWhenDisabled>` and `<Button disabled>` don't know about each other, so the button just renders as a regular `disabled` button.

I also tried the following with no luck (it renders with both `disabled` and `aria-disabled`):

```jsx
<Ariakit.MenuButton
  accessibleWhenDisabled
  disabled
  render={<Ariakit.Button />}
/>
```

The closest I got was using this:

```jsx
<Ariakit.MenuButton
  render={<Ariakit.Button accessibleWhenDisabled disabled />}
/>
```

However, that also does not work fully. The button renders with `aria-disabled`, and can be focused, and prevents the menu from being opened through clicks, <kbd>Enter</kbd> and <kbd>Spacebar</kbd>; **but** the menu still opens when pressing arrow-down or arrow-up.

See stackblitz repro: https://stackblitz.com/edit/xw2vog?file=menu%2Findex.jsx

I'm not opening a full bug report, because I'm not sure if this is a bug/edge-case or if this behavior is intentional due to React's one-way data-flow. Any thoughts?

**Comment by diegohaz** [comment]
Maybe this is not a bug, but we definitely want to support all those combinations.

https://github.com/ariakit/ariakit/issues/4133

**Reply by mayank99** [reply]
A good first step could be to include `accessibleWhenDisabled` in the list of props that get merged (Right now it looks like just HTML props, ref, and event handlers, based on the [composition guide](https://ariakit.org/guide/composition#merging-the-rendered-element-props)). This is likely a broader issue with `Role`, rather than specific to `MenuButton`.

**Reply by diegohaz** [reply]
I'm not sure we can do that consistently. You can always have a custom `Button` component that abstracts `<Ariakit.Button accessibleWhenDisabled>`. In this case, this prop wouldn't be seen by the `render` prop logic.

We can only trust props that are actually passed to the underlying DOM element. That could be either a data attribute or our custom `useMetadataProps` hook:

https://github.com/ariakit/ariakit/blob/4acdb60379a8cf80949bb882c99fe438ac36cf34/packages/ariakit-react-core/src/utils/hooks.ts#L366-L380
